PROTOCOL.md

```diff
@@ -2,179 +2,287 @@
-the read operation MUST yield EOF.
-
-* A close operation by the application on the a byte-stream field MUST trigger the closing of the writing end
-of the corresponding SPDY stream.
-
-#### Optional method: file descriptor passing
+* When decoding a message including 1 or more byte stream values, the receiver
+MUST store the unique identifier of each value in a session-wide table of
+pending byte streams. It MAY then immediately pass the decoded message to the
+application.
 
-*FIXME*
+* The sender SHOULD cap the size of its pending byte streams table to a
+reasonable value. It MAY make that value configurable by the application. If it
+receives a message with 1 or more byte stream references, and the table
+is full, the sender MAY suspend processing of the message until there is room in
+the table.
 
-### Sending nested channels
+* When receiving new streams which include the header key "*libchan-ref*", the
+receiver MUST lookup that header value in the table of pending byte streams. If
+the value is registered in the table, that stream MUST be passed to the
+application.
 
-*FIXME*
+On either end, once the stream for a byte-stream value is established, it MUST
+be exposed to the application as follows:
+
+* After sending each of those streams, each write operation by the application
+to a byte-stream field MUST trigger the sending of a single data frame on the
+corresponding stream.
+
+* Each read operation by the application from a byte-stream field MUST yield
+the content of the next data frame received on the corresponding stream. If the
+reading end of the stream is closed, the read operation MUST yield EOF.
+
+* A close operation by the application on the a byte-stream field MUST trigger
+the closing of the writing end of the corresponding SPDY stream.
+
+## Message encoding
+A message may be any type which supported by the libchan encoder.  A libchan
+message encoder must support encoding raw byte stream types as well as channels.
+In addition to the libchan data types, time must also be encoded as a custom
+type to increase portability of the protocol.
+
+Currently supported message encoders are msgpack and soon CBOR.
+
+### Custom Types
+Each custom type defines a type code and the byte layout to represent that type.
+Directions of descriptions are from the point of view of the endpoint encoding.
+All multi-byte integers are encoded big endian. The length of bytes of the
+encoded value will be provided by the encoding format, allowing integer values
+to be variable length.
+
+| Type | Code | Byte Layout|
+|---|---|---|
+| Duplex Byte Stream | 1 | 4 or 8 byte integer identifier |
+| Inbound Byte Stream | 2 | 4 or 8 byte integer identifier |
+| Outbound Byte Stream | 3 | 4 or 8 byte integer identifier |
+| Inbound channel | 4 | 4 or 8 byte integer identifier |
+| Outbound channel | 5 | 4 or 8 byte integer identifier |
+| time | 6 | 8 byte integer seconds + 4 byte integer nanoseconds  |
```

Comment:
I think have variable integer identifier is a bad idea, as they are impossible to parse in msgpack (you would need one more byte for header or some weird trick). 8 byte or 4 byte are both fine, just stick to one.